### DISCUSSION TITLE
Seeing what looks like a memory leak

### DISCUSSION THREAD

**Discussion by cpoppema** [emmett-framework/granian#268] [General]
For two projects I replaced uvicorn with uvloop with granian and my monitoring tools show an increase in memory usage. I am wondering if this is something I have to accept with granian - I don't know if it'll flatten out at some point - or if there is an underlying issue I might be able to help debug. Benchmarking is all about request numbers but does not mention any memory increase, so I suspect an issue.

Figures below are from 2 services running fastapi + pydantic + sqlalchemy.

service 1, graph is from 7 days:

cpu before (uvicorn)/after (granian)
![2024-04-12-1712919886_screenshot](https://github.com/emmett-framework/granian/assets/1432901/dcf5376d-47ad-4c82-8c84-4bcb7b42a6b7)

requests before (uvicorn)/after (granian)
![2024-04-12-1712919843_screenshot](https://github.com/emmett-framework/granian/assets/1432901/d38b2644-5e5b-4ed1-bc22-922230abae8d)

memory before (uvicorn)/after (granian)
![2024-04-12-1712919806_screenshot](https://github.com/emmett-framework/granian/assets/1432901/6a9f56a7-5144-482b-af95-5ccecc491520)

service 2, graph is from 2 days:

cpu before (uvicorn)/after (granian)
![2024-04-12-1712920252_screenshot](https://github.com/emmett-framework/granian/assets/1432901/e2dfed36-a4f8-49a0-bdad-e20a4fb9c018)

requests before (uvicorn)/after (granian)
![2024-04-12-1712920168_screenshot](https://github.com/emmett-framework/granian/assets/1432901/585c2933-5880-4627-a70c-7249829748fc)

memory before (uvicorn)/after (granian)
![2024-04-12-1712920305_screenshot](https://github.com/emmett-framework/granian/assets/1432901/29697785-c24e-45ce-887e-15d64d437ee0)

**Comment by gi0baro** [comment]
My best guess at the moment is that might be related to `mimalloc` allocator.
Is that Linux on an `x86_64` architecture?

**Reply by cpoppema** [reply]
Yes indeed.

```
$ uname -a
Linux xxxxxxxxxxxx x.y.z-a.b.amzn2.x86_64 #1 SMP Mon Apr 24 23:34:06 UTC 2023 x86_64 GNU/Linux
```

**Reply by gi0baro** [reply]
👍 I gonna run some tests with allocators and see if anything can be improved.

Are you able to keep at least one of those 2 services running with Granian to see if it flatten out at a certain point in time/value?

**Reply by cpoppema** [reply]
Yes I should be able to for at least one of them :sweat_smile: - at some point I know some agent process will terminate it/restart the container if it exceeds its memory limits which is currently set at 512M. Service 1 had already reached the first 80% alert threshold after running for 26 hours and I had to rollback to uvicorn for now.

**Reply by cpoppema** [reply]
There was a lot less traffic over the weekend, but memory usage did not stabilize. This morning the OOMKiller paid a visit.

![2024-04-15-1713162520_screenshot](https://github.com/emmett-framework/granian/assets/1432901/4c9262b6-a400-47f6-a07c-1d035b4f9441)

edit: I realize I cut of the y-axis, but it is killed at exactly 488MB (512Mi)

**Reply by gi0baro** [reply]
@cpoppema thank you for the update, current master is now using jemalloc on linux; I still need to perform some tests but I expect it to land in 1.2.3 in the following days.
You might also try to build granian from `master` branch and already check if anything changes on your workloads.

**Reply by cpoppema** [reply]
Thanks. I will give building from master a try!

**Reply by cpoppema** [reply]
I build the .so file and dropped it in my site-packages, but because I don't see a difference I wonder if I have done it correctly. Running `granian --version` _does_ report 1.2.3.

I've kept 1.2.2 in my requirements.txt and after installing those I am running the following:

```
curl --proto '=https' --tlsv1.2 https://sh.rustup.rs -sSf | sh -s -- -y && \
    cd /tmp && \
    curl https://github.com/emmett-framework/granian/archive/refs/heads/master.zip -Lo master.zip && \
    unzip master.zip && \
    cd granian-master && \
    ~/.cargo/bin/cargo build --release && \
    cp target/release/lib_granian.so /usr/local/lib/python3.11/site-packages/granian/_granian.cpython-311-x86_64-linux-gnu.so && \
    cd /app && \
    rm -rf /tmp/granian-master ~/.cargo
```

![2024-04-16-1713251863_screenshot](https://github.com/emmett-framework/granian/assets/1432901/84ab2c5e-e8bf-421d-bee7-b9e2a50bf0de)


This is from service 1 which I reverted to uvicorn before.
- yellow and red, the last 24 hours: yellow=uvicorn, red=granian 1.2.3
- blue: an offset from the last 24 hours when running uvicorn
- green: an offset from 96 hours ago when running granian 1.2.2

**Reply by cpoppema** [reply]
This morning I overlooked the official 1.2.3 release, so I will redeploy with your release to see if it makes a difference. :crossed_fingers: The memory from my manual build so far looks like this:

![2024-04-16-1713261754_screenshot](https://github.com/emmett-framework/granian/assets/1432901/16a2afc1-5705-4c3f-9dcc-5b39261e2196)

**Reply by gi0baro** [reply]
@cpoppema thank you for the updates.

I doubt official 1.2.3 will make any difference, it seems there's definitely something going on here.
Probably the code is missing some Python GC calls (https://pyo3.rs/main/class/protocols#garbage-collector-integration), but given https://github.com/emmett-framework/granian/pull/265 is changing a lot of things, I need to schedule memory allocation tests after that is merged, and thus this will probably see some enhancements with 1.3.x series, which I expect to release in the next couple of weeks.
Thank you for your patience.

**Reply by cpoppema** [reply]
No problem and you're right, it makes no difference (orange=version 1.2.3 from pypi).

![2024-04-16-1713270547_screenshot](https://github.com/emmett-framework/granian/assets/1432901/1dadb77d-4255-4fc5-b835-63f3afb7e524)

Feel free to @ me if want me to test a new version. (In the meantime I will most likely switch to hypercorn as a replacement for uvicorn)

**Reply by gi0baro** [reply]
@cpoppema can you provide additional details on how do you run Granian on such services? eg: number of workers, threads, blocking threads, threading mode..

**Reply by cpoppema** [reply]
I have been running with the following command:
```
granian --interface asgi service.server:app --host=0.0.0.0 --loop uvloop
```
